chain/src/pipe.rs

```diff
@@ -276,6 +276,45 @@ pub fn sync_block_headers(
 	}
 }
 
+/// Process the kernels.
+/// This is only ever used during sync.
+pub fn sync_kernels(
+	first_kernel_index: u64,
+	kernels: &Vec<TxKernel>,
+	ctx: &mut BlockContext,
+) -> Result<(), Error> {
+	if let Some(kernel) = kernels.first() {
+		debug!(
+			"pipe: sync_kernels: {} kernels from {} at {}",
+			kernels.len(),
+			kernel.hash(),
+			first_kernel_index
+		);
+	} else {
+		return Ok(());
+	}
+
+	if ctx.txhashset.num_kernels() < (first_kernel_index + 1) {
+		txhashset::extending(&mut ctx.txhashset, &mut ctx.batch, |extension| {
+			// DAVID: Rewind mmr to correct kernel index just to be safe?
+
+			for kernel in kernels {
+				// Ensure kernel is self-consistent
+				kernel.verify()?;
+
+				// Apply the kernel to the kernel MMR.
+				extension.apply_kernel(kernel)?;
+
+				// DAVID: If kernel is last in block, validate root.
+				// Use RewindableKernelView? Or modify validate_kernel_history?
+			}
```

Comment:
I'd consider pushing that whole loop down in txhashset in some `apply_kernels`.

Reply:
My bad, looks like apply does it. But there should be some clearer separation around validation. Here you're verifying each kernel, but the whole state validation will do that again. So we either slip the latter one or this one.

Reply:
Depends what you mean by whole state validation. If you're talking about after the TxHashSet is fully downloaded and processed, sure. But then we'd have to go back and start the kernel validation process all over again. If instead you're referring to the expected call to "validate_kernel_history" on line 309, then I disagree. It doesn't verify the signature of the kernel. It only verifies that the MMR root matches the header.

Reply:
After re-reading, it appears you're talking about verify_kernel_signatures in Extension. Yea, I can probably skip that verify if it's already been done. It'd probably be easiest to store a 'validated' boolean on the kernel. If you're opposed to adding the extra byte (since kernels are all written to disk), I can look into other options.